### ISSUE TITLE
Investigate options for replacing sqlite with postgres

### ISSUE THREAD

**Issue by dnephin** [infrahq/infra#2584]
## Background

Today the Infra server supports both sqlite and postgres as database engines. Most of that compatibility is provided by Gorm, but some of it (`handleError`, and any queries we have to build ourselves, ex: #2578) is handled in the `data` package.

We'd like to only support postgres in production for a number of reasons (reduce our maintenance burden, we don't have any way to support multiple server replicas with sqlite). We'd also like to replace Gorm (#2415), which may slightly increase the maintenance cost of multiple databases.

## Problem

Today we run all `data` tests against both databases, but we run all other tests against only sqlite. We can see from the test ouput ([log](https://github.com/infrahq/infra/runs/7343470624?check_suite_focus=true), look for `/postgres` and `/sqlite` ) that running tests against postgres is 5x slower in CI with `-race` enabled (on my local setup its 30x slower).

Running all of our tests against only postgres would significantly slow down our test suite, which would decrease productivity.

It seems like we have three options:
1. Continue to pay the small maintenance cost of making queries work with sqlite, so that we have a convenient way of testing other layers of our application, and so that we have a convenient workflow for trying out the rest of Infra
2. Completely remove sqlite support, and find some other way to provide a faster data layer for testing other parts of the system (API, CLI, core business logic). For example, a full in-memory fake implementation.
3. Spent time investigating why the postgres version is so much slower, and find ways to optimize it. My guess is that it may have something to do with gorm and/or migrations.

It's also important to note that today we can't easily test all our of DB queries in the data layer because of the `SelectorFunc` pattern. We plan on replacing selector funcs with a struct of input values, so that all query construction happens in a single function in `data`. That change should make it much easier to test all queries against postgres.

**Comment by dnephin** [comment]
Given these trade-offs, my preference would be to keep sqlite for dev and demo purposes. We may have to write a second very of some queries, but we don't really need to care much about the performance of those queries.

If we compare that to the alternatives, I think it's likely less work to write a similar sqlite compatible query (for the few rare cases where we can't use a single query for both), than to try and create a completely separate implementation of our data layer.

**Comment by dnephin** [comment]
I tried most of the suggestions from this [SO answer](https://stackoverflow.com/questions/9407442/optimise-postgresql-for-fast-testing/9407940#9407940) (using a tmpfs, `fsync=off`, `full_page_writes=off`) and I see almost no difference in the timings.  I can see the command line options in the output of `ps aux`, so I think they are being applied correctly.

I was not able to try `UNLOGGED` yet because that requires changing the `CREATE TABLE` expression, and I'm not sure if there is an easy way to do that with gorm.

**Comment by dnephin** [comment]
I added some regions to the test setup using [runtime/trace](https://pkg.go.dev/runtime/trace), I pushed those changes here: https://github.com/infrahq/infra/compare/dnephin/db-postgres-trace

I ran the tests with `-trace=trace.out`, and saw this:

![trace](https://user-images.githubusercontent.com/442180/179839606-35ae2fed-21b4-4ecf-b060-a91c5f3e19fd.png)

In this environment an sqlite test runs in about 10ms, and a postgres test runs in about 350ms.

* creating the driver takes very little time (expected)
* opening the connection takes about 3ms of network wait (and another ~2.5ms of execution), still pretty negligible and likely some of that cost is in gorm.
* creating the schema takes very little time (0.5ms)
* dropping the schema takes 25-40ms per test, all of which is network wait. This accounts for some of the extra time
* most of the extra time is in `setupDB`, which takes 10ms with sqlite, and 250-350ms with postgres, most of which is in network wait

Looking at the "network blocking" profile (called "network wait" in the other view), it seems most of the time is a result of running the migrations. I think this confirms my initial hypothesis.

I suspect that by replacing gormigrate we'll find a way to load the schema much faster, which should bring the runtime of postgres tests closer to the tests running sqlite.
